Sming/Services/SpifFS/spiffs_sming.c

```diff
@@ -64,10 +64,10 @@ bool spiffs_format_internal(spiffs_config *cfg)
   sect_first = cfg->phys_addr;
   sect_first = flashmem_get_sector_of_address(sect_first);
   sect_last = cfg->phys_addr + cfg->phys_size;
-  sect_last = flashmem_get_sector_of_address(sect_last);
+  sect_last = flashmem_get_sector_of_address(sect_last) - 1;
```

Comment:
I think the correct code would be:
```
  sect_last = cfg->phys_addr + cfg->phys_size - 1;
  sect_last = flashmem_get_sector_of_address(sect_last);
```

Reply:
The first value assigned to `sect_last` is the physical address of the first byte which does not belong the the SpifFS. If `cfg->phys_addr + cfg->phys_size` is a multiple of `INTERNAL_FLASH_SECTOR_SIZE` (which is 4096), then it would make no difference but it would be logically wrong. 
Assume `phys_addr+phys_size` would not be a multiple of `INTERNAL_FLASH_SECTOR_SIZE` (e.g. `INTERNAL_FLASH_SECTOR_SIZE/2`). Then you subtract one byte from `sect_last` - now `sect_last` points to the last byte which belongs to SpifFS. Now `flashmem_get_sector_of_address()` calculates the sector as follows: `address / INTERNAL_FLASH_SECTOR_SIZE`. The resulting sector is the same as if you did not subtract 1 before (because of the division of integers). In this case you would override/delete the data from the address `cfg->phys_addr + cfg->phys_size` to the end of the sector.
If you subtract one from `sect_last` after the "second step" (calculating the sector) you will not override other data.

Reply:
I like good discussions especially when they lead to results. I still think that Mike is right here. I will post below my reasoning and you can follow up with your arguments.

Let's have the following setup:
Sector size: 5
Start address: 4
Size: 2

![image](https://user-images.githubusercontent.com/247452/55060339-8b4b3880-5071-11e9-8341-d70da32b704a.png)

Mike's calculation:
```
sect_last = cfg->phys_addr + cfg->phys_size - 1;  // 4 + 2 - 1 = 5 <-- last byte inclusive is the 5th one
sect_last = flashmem_get_sector_of_address(sect_last); // 5/5 = 1  <-- the sector for the 5th byte is 1.
```
Which means that sector 0 and sector 1 have to be erased.  Which IMHO is correct. 

Now your solution suggest the following
```
sect_last = cfg->phys_addr + cfg->phys_size; // 4 + 2 = 6
sect_last = flashmem_get_sector_of_address(sect_last) - 1; // (6/5) - 1 = 1 -1 =  0  <-- the sector that has to be erased is 0
```

Which means that only sector 0 has to be erased. But that cannot be right because byte 5 has to be written to sector 1, which will not be erased here. Am I missing something here?

Reply:
No you are not missing something. But I think it's better to not use a block instead of deleting other data (here bytes 6,7,8,9). Another thing I didn't consider is the same aspect but for `cfg->phys_addr` (so also bytes 0,1,2,3 are deleted). In your example should no block be erased, because both blocks may contain other data which may not be erased. I think it would be reasonable to check this in `overrides.c`:

```
#ifdef RBOOT_SPIFFS_0
    #if (((RBOOT_SPIFFS_0) % (INTERNAL_FLASH_SECTOR_SIZE)) != 0)
        #error "RBOOT_SPIFFS_0 has to be a multiple of INTERNAL_FLASH_SECTOR_SIZE"  // or warning
    #endif
#endif

#ifdef RBOOT_SPIFFS_1
    #if (((RBOOT_SPIFFS_1) % (INTERNAL_FLASH_SECTOR_SIZE)) != 0)
        #error "RBOOT_SPIFFS_0 has to be a multiple of INTERNAL_FLASH_SECTOR_SIZE"  // or warning
    #endif
#endif

#if (((SPIFF_SIZE) % (INTERNAL_FLASH_SECTOR_SIZE)) != 0)
    #error "SPIFF_SIZE has to be a multiple of INTERNAL_FLASH_SECTOR_SIZE"  // or warning
#endif
```

When the developer tries using values which will delete other data it prevents him from doing this or just get warned and does not have to search for the problem.